xds/src/main/java/io/grpc/xds/LeastRequestLoadBalancer.java

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.SHUTDOWN;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.LeastRequestLoadBalancerProvider.DEFAULT_CHOICE_COUNT;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import io.grpc.Attributes;
+import io.grpc.ClientStreamTracer;
+import io.grpc.ClientStreamTracer.StreamInfo;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.xds.ThreadSafeRandom.ThreadSafeRandomImpl;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link LoadBalancer} that provides least request load balancing based on
+ * outstanding request counters.
+ *
+ */
+public class LeastRequestLoadBalancer extends LoadBalancer {
```

Comment:
@ejona86 I'm not finished with this yet but made a draft PR for early input.
This class in particular is copying a very large amount of code from `RoundRobinLoadBalancer` .
There are a couple ways to approach this that I can think of:
1) Live with duplicated code/tests
2) Break out common code into an `AbstractLoadBalancer` or similar.
    Such a class would need to be public unless it makes more sense to move `least_request` to grpc-core instead. This might be fine since it's not tied to xDS in any particular way and can be used standalone.

Reply:
I count around 100 lines (raw; including javadoc, etc) that could be shared with round_robin, although there is custom logic that needs to be mixed in with those, so the actual gains will be less. The diffing logic in `handleResolvedAddresses()` seems the strongest candidate to move into a "subchannel collection" util and it could help to some degree with `updateBalancingState()` as well.

But I don't want to get too side-tracked with that. We are talking about just two implementations being able to share code, and they are in different modules.

I think for this PR we should just accept the duplication, and let any utility-creating work happen afterward (if ever). 

Both xds and RR can access things in io.grpc.internal, so it could be an internal API for a while to see how well it goes. I'm against an AbstractLoadBalancer in the normal sense; this logic is actually specialized to RR-like policies. I'd much rather have helper utility classes/methods and avoid them being NameResolvers themselves, as `extends` is poorly suited for sharing behavior.

> unless it makes more sense to move least_request to grpc-core instead

I'm not quite sure where we'd want to put it. I really want core to be bare-bones stuff. I agree the code isn't really xds-specific, although as defined in the gRFC it is xds-specific as the long-term name is `least_request_experimental`. I think we'd figure out a different home if it ever becomes `least_request`.

Reply:
> in the gRFC it is xds-specific as the long-term name is least_request_experimental. I think we'd figure out a different home if it ever becomes least_request.

One thing I noticed first today was that the gRFC for ring hash seemed to use a mix of `ring_hash_experimental` and `ring_hash` for the policy name (see [here](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md#ring_hash_experimental-lb-policy)).
In the end, it was implemented as `ring_hash` in #7943 . I don't have any preference since it will work with xDS regardless of the internal name but right now I've named it `least_request` for consistency with `ring_hash`. Do you want me to name it `least_request_experimental` here to be consistent with the gRFC instead?

Also if you are fine with duplicated code/tests, it seems like the most appropriate home for this now at least is the xds module.
I'll go ahead and add the final tests then and not bother with breaking things out of round robin.

Reply:
Alright, I'll update this PR with the correct `least_request_experimental` name tomorrow when I have time.
Besides that, this is pretty much ready for review when you have time for it.

For review of LeastRequestLoadBalancer(Test) I suggest comparing it with the current RoundRobinLoadBalancer(Test) classes since there's a fair bit of copied code.
I've also tested the snapshot build of this PR in a more realistic setup to validate that it works as intended.